dotnet/src/SemanticKernel/Kernel.cs

```diff
@@ -53,6 +54,12 @@ public sealed class Kernel : IKernel, IDisposable
     /// </summary>
     public static KernelBuilder Builder => new();
 
+    /// <inheritdoc/>
+    public event EventHandler<FunctionInvokingEventArgs>? FunctionInvoking;
```

Comment:
Can we use `AsyncEventHander`?... 
```
public delegate Task AsyncEventHandler<TEventArgs>(
    object? sender,
    TEventArgs args,
    CancellationToken cancellationToken = default) where TEventArgs : EventArgs;
```

That way, handlers can naturally be written as awaitable Task methods. Then, when invoked, each callback is awaited. (it's likely that many handlers won't need to be awaitable, but it's far easier/cleaner to implement an async-over-sync handler than a sync-over-async method).

`public event AsyncEventHandler<FunctionInvokingEventArgs>? FunctionInvoking;`

Reply:
There is no AsyncEventHandler supported by dotnet.

One hacky way would be something like Example below:
 
```
public event Func<object, EventArgs, Task>? FunctionInvoking;
``` 
https://codereview.stackexchange.com/questions/261396/asynchronous-event-handler

Keep note that trying this approach got [CA1003 Warning](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1003)